File: scripts/import_inventories.go
//go:build ignore

package main

import (
	"encoding/json"
	"os"

	"github.com/uwdata/mosaic/packages/server/duckdb-server-go/pkg/functionset"
)

func main() {
	extensions := map[string]any{}
	for _, extension := range functionset.CoreExtensions() {
		extensions[string(extension)] = map[string]any{"compute": extension.Compute(), "elevated": extension.Elevated()}
	}
	if err := json.NewEncoder(os.Stdout).Encode(map[string]any{
		"duckdb_version": "1.5.5",
		"defaults": functionset.DefaultFunctions(),
		"core": map[string]any{
			"operators": functionset.BuiltinOperators(), "aggregates": functionset.BuiltinAggregates(),
			"windows": functionset.BuiltinWindows(), "syntax_helpers": functionset.BuiltinSyntaxHelpers(),
			"table_generators": functionset.BuiltinTableGenerators(), "scalars": functionset.CommonScalarFunctions(),
		},
		"extensions": extensions,
	}); err != nil {
		panic(err)
	}
}

File: scripts/import_inventories.py
"""One-time migration from the reviewed Mosaic inventories; not part of the build."""
import json
from pathlib import Path
import subprocess
import sys

root = Path(__file__).resolve().parents[1]
data = subprocess.check_output(
    ["go", "run", str(root / "scripts/import_inventories.go")], cwd=sys.argv[1], text=True
)
(root / "inventories").mkdir(exist_ok=True)
data = json.loads(data)
source = Path(sys.argv[1]) / "pkg/functionset"
provenance = "https://github.com/uwdata/mosaic/tree/3eb74ea8/packages/server/duckdb-server-go/pkg/functionset"
core = {
    "name": "core", "reviewed_duckdb": "1.5.5", "source": provenance + "/functionset.go",
    "notes": "Reviewed defaults migrated from Mosaic. Unlisted names remain excluded pending review.",
    "groups": data["core"], "compute": sorted(set(n for group in data["core"].values() for n in group)),
    "elevated": [],
}
(root / "inventories/core.json").write_text(json.dumps(core, indent=2) + "\n")
(root / "inventories/extensions").mkdir(exist_ok=True)
for name, groups in data["extensions"].items():
    comments = [line.strip()[3:] for line in (source / (name + ".go")).read_text().splitlines() if line.strip().startswith("// ")]
    entry = {"name": name, "reviewed_duckdb": "1.5.5", "source": provenance + "/" + name + ".go",
             "notes": comments, **groups}
    (root / "inventories/extensions" / (name + ".json")).write_text(json.dumps(entry, indent=2) + "\n")
